    We will do
    1. Fit a logistic regression model to only one predictor and understand how to interpret the results.
    2. Fit the logistic regression model to the entire dataset.
    3. Take the results from the logistic regression model to create a bootstrap with confidence intervals.

In [1]:
import sys
import numpy as np
import pandas as pd
pd.set_option('display.max_rows', 999)
pd.set_option('display.max_columns', 999)
pd.set_option('display.width', 1000)
pd.set_option('display.notebook_repr_html', True)
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from statsmodels.api import OLS
from statsmodels.api import add_constant
from statsmodels.regression.linear_model import RegressionResults
from sklearn.model_selection import KFold
from sklearn.linear_model import LinearRegression
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import r2_score
from collections import Counter
sns.set(style="ticks")
%matplotlib inline

import warnings
warnings.filterwarnings("ignore")

matplotlib.rcParams['figure.figsize'] = (13.0, 6.0)


#### Loading Data

In [2]:
df = pd.read_csv('/cancer_data_log.csv')
df.head()

Y    Gene 1    Gene 2    Gene 3    Gene 4    Gene 5    Gene 6    Gene 7    Gene 8    Gene 9   Gene 10   Gene 11   Gene 12   Gene 13   Gene 14   Gene 15   Gene 16   Gene 17   Gene 18   Gene 19   Gene 20   Gene 21   Gene 22   Gene 23   Gene 24   Gene 25   Gene 26   Gene 27   Gene 28   Gene 29   Gene 30   Gene 31   Gene 32   Gene 33   Gene 34   Gene 35   Gene 36   Gene 37   Gene 38   Gene 39   Gene 40   Gene 41   Gene 42   Gene 43   Gene 44   Gene 45   Gene 46   Gene 47   Gene 48   Gene 49   Gene 50   Gene 51   Gene 52   Gene 53   Gene 54   Gene 55   Gene 56   Gene 57   Gene 58   Gene 59   Gene 60   Gene 61   Gene 62   Gene 63   Gene 64   Gene 65   Gene 66   Gene 67   Gene 68   Gene 69   Gene 70   Gene 71   Gene 72   Gene 73   Gene 74   Gene 75   Gene 76   Gene 77   Gene 78   Gene 79   Gene 80   Gene 81   Gene 82   Gene 83   Gene 84   Gene 85   Gene 86   Gene 87   Gene 88   Gene 89   Gene 90   Gene 91   Gene 92   Gene 93   Gene 94   Gene 95   Gene 96   Gene 97   Gene 98   Gene 99  \
0  0 -0.927178 -0.740039 -0.532016 -1.097891 -0.986673  0.018752 -0.863326  3.959997 -0.398219 -1.097891  0.183420 -1.097891 -1.097891  0.301076 -1.097891  2.172958  0.606712  1.420045  3.485251 -1.097891 -1.097891 -1.097891  1.223614  1.421553  0.732994 -0.871084 -1.097891 -0.087447 -0.087447 -1.097891  1.716598 -1.097891 -0.351254 -1.097891  1.110441 -0.686979  2.552726 -1.097891 -1.097891 -1.022364 -1.097891 -1.097891 -1.097891  0.585757 -0.803376 -1.097891 -1.097891 -0.618410  0.196629 -1.097891 -1.097891 -0.588759 -1.097891  2.747928  0.749564 -0.417653  1.417023 -1.097891  1.835157 -0.840413 -0.803376 -1.097891  0.043598 -0.160175 -0.422572 -1.097891 -0.894733 -1.097891 -0.412758 -0.202598  2.107798 -1.097891 -0.403041  0.630608  0.128733 -0.190845  0.382527 -0.202598  0.604983  0.708381  0.950157  3.432058 -1.097891 -0.796133 -0.259388 -1.097891  1.357764 -1.097891 -1.097891 -1.097891 -1.097891 -0.333080  0.058813  0.082678 -0.977950  0.777542 -0.293378  0.439465 -1.097891   
1  0 -0.835899 -0.835899 -0.585647 -0.835899 -0.329768 -0.835899 -0.835899  4.037788 -0.835899 -0.835899  0.934701 -0.835899 -0.835899  0.437471  2.026290  2.828351  1.291165  2.872351  3.434738 -0.835899 -0.835899 -0.835899 -0.174752 -0.744989 -0.514960 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899  0.111498 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899  2.981215 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.514960 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899  3.243331 -0.661994 -0.835899  0.146187 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899  0.075500 -0.835899 -0.835899 -0.835899 -0.835899 -0.835899  0.243271 -0.835899 -0.275248 -0.835899  1.526308 -0.835899 -0.835899 -0.835899 -0.835899  0.146187 -0.835899  3.180358 -0.835899 -0.835899 -0.835899 -0.835899  2.375943 -0.835899 -0.661994 -0.835899 -0.835899 -0.835899 -0.449153 -0.835899 -0.835899  1.714795 -0.835899  1.159392 -0.835899   
2  0  0.236073  0.252645 -1.154351 -0.372372 -0.338900  1.253467  0.506325  3.868961  0.201932 -1.154351 -1.073012  0.148014 -1.154351  0.576509 -1.154351  0.892058  0.973397 -0.930445  1.619579 -0.245707  0.456109  0.007132  1.331104  0.852357  0.966315  0.049812 -0.372372  0.598677  0.576509 -0.443531  2.016960  0.443080  0.682003 -0.443531  1.402263 -1.154351 -0.275655 -1.154351  0.070377  0.110078  0.007132  0.201932 -1.154351  1.034625 -0.481467 -1.154351 -1.154351 -0.701503  0.884264  0.553749 -0.481467 -0.061178 -1.154351 -1.154351  1.258531  0.345838  2.016960 -0.998755  0.252645 -0.061178  0.494045 -1.154351  0.148014 -0.753241  0.284853 -1.154351 -1.154351  0.166368  0.530367 -0.481467  2.362299 -1.154351  0.876398  1.243246 -1.154351  0.620283  1.410755  0.184335  0.219173  1.431622  1.475809  2.739170 -0.753241  0.236073  0.049812 -0.652723  0.028738 -1.154351  0.711266 -1.154351 -1.154351  0.506325 -1.154351  0.827618  0.028738 -1.154351  0.360426 -1.154351 -1.154351   
3  0 -0.748

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102 entries, 0 to 101
Columns: 6034 entries, Y to Gene 6033
dtypes: float64(6033), int64(1)
memory usage: 4.7 MB


In [4]:
df.describe()

Y      Gene 1      Gene 2      Gene 3      Gene 4      Gene 5      Gene 6      Gene 7      Gene 8      Gene 9     Gene 10     Gene 11     Gene 12     Gene 13     Gene 14     Gene 15     Gene 16     Gene 17     Gene 18     Gene 19     Gene 20     Gene 21     Gene 22     Gene 23     Gene 24     Gene 25     Gene 26     Gene 27     Gene 28     Gene 29     Gene 30     Gene 31     Gene 32     Gene 33     Gene 34     Gene 35     Gene 36     Gene 37     Gene 38     Gene 39     Gene 40     Gene 41     Gene 42     Gene 43     Gene 44     Gene 45     Gene 46     Gene 47     Gene 48     Gene 49     Gene 50     Gene 51     Gene 52     Gene 53     Gene 54     Gene 55     Gene 56     Gene 57     Gene 58     Gene 59     Gene 60     Gene 61     Gene 62     Gene 63     Gene 64     Gene 65     Gene 66     Gene 67     Gene 68     Gene 69     Gene 70     Gene 71     Gene 72     Gene 73     Gene 74     Gene 75     Gene 76     Gene 77     Gene 78     Gene 79     Gene 80     Gene 81  \
count  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000  102.000000   
mean     0.509804   -0.623086   -0.665809   -0.586693   -0.895592   -0.438002   -0.152997   -0.593764    4.361345   -0.553667   -0.951837    0.210988   -0.655313   -0.942147    1.056787    1.048407    2.420894    1.385449    1.984182    3.279895   -0.804082   -0.663843   -0.547087    0.982289    0.420177    0.589138   -0.654416   -0.763602   -0.177263   -0.354601   -0.733137    1.019639   -0.853698   -0.387593   -0.723554    0.333692   -0.870300    2.302401   -0.923628   -0.615913   -0.573766   -0.665838   -0.520763   -0.880360    0.206195   -0.512913   -0.927772   -0.818706   -0.693632   -0.077625   -0.674515   -0.792538   -0.596294   -0.943320    2.282409    0.475842   -0.598812    1.027012   -0.725107    0.342511   -0.816244   -0.278585   -0.498915   -0.177066   -0.528816   -0.262572   -0.948803   -0.852756   -0.833931   -0.218727   -0.653240    1.429048   -0.771873   -0.073120    0.287660    0.001412   -0.551145   -0.686902   -0.494723   -0.368184    0.538564    0.488985   
std      0.502373    0.409905    0.378207    0.497839    0.190989    0.597182    0.789601    0.414790    0.410420    0.399223    0.181319    0.606445    0.469975    0.243418    0.730810    1.528193    0.703898    0.712412    1.423613    0.860988    0.314300    0.455784    0.569803    0.607126    0.735268    0.669660    0.401457    0.292917    0.518869    0.560968    0.427445    0.629158    0.282037    0.527495    0.362698    0.837500    0.241137    1.421855    0.202965    0.667866    0.430846    0.352903    0.511037    0.225416    0.577369    0.537635    0.211668    0.339570    0.372069    0.664546    0.484781    0.311698    0.368384    0.191362    1.739226    0.716093    0.408567    0.581369    0.322546    1.304549    0.285212    0.568969    0.569240    0.597381    0.470440    0.765572    0.203211    0.264404    0.293816    0.563874    0.391422    0.642131    0.350769    0.536137    0.673359    0.957940    0.498118    0.460433    0.428367    0.638959    0.628258    0.855911   
min      0.000000   -1.130098   -1.206937   

`Y` here is the response variable (`Y` = 1 is cancer and `Y` = 0 is no cancer) and the 6033 predictor variables are gene expressions.Each row represents a person.

In [9]:
#Tarin test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( df.drop('Y',axis = 1), df['Y'], test_size=0.33, random_state=42)

In [11]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68 entries, 28 to 51
Columns: 6033 entries, Gene 1 to Gene 6033
dtypes: float64(6033)
memory usage: 3.1 MB


In [14]:
y_train.shape

(68L,)

In [16]:
#Normalization
X_train_n = (X_train - X_train.min(axis=0))/(X_train.max(axis=0) - X_train.min(axis=0))
X_test_n = (X_test - X_train.min(axis=0))/(X_train.max(axis=0) - X_train.min(axis=0))

#### Using one single predictor to fit a Logistic Regression Model  
Using Gene 10

In [51]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression(C=1000000, fit_intercept=True) #The C value here is a way to prevent regularization
X_train_gene10 = X_train["Gene 10"]
X_test_gene10 = X_test["Gene 10"]
log_reg.fit(X_train_gene10.reshape(-1,1), y_train)
y_predict_train_log = log_reg.predict(X_train_gene10.reshape(-1,1))
y_predict_test_log = log_reg.predict(X_test_gene10.reshape(-1,1))
print "Training Accuracy for Logistic Regression: ", accuracy_score(y_train, y_predict_train_log)
print "Test Accuracy for Logistic Regression: ", accuracy_score(y_test, y_predict_test_log)

Training Accuracy for Logistic Regression:  0.632352941176
Test Accuracy for Logistic Regression:  0.529411764706


In [52]:
log_reg.intercept_

array([ 1.18325051])

In [56]:
log_reg.coef_

array([[ 1.05701323]])

Calculating the probability of cancer in our training set  without using predict_proba() 

$$ p = \cfrac{e^\eta}{1+e^\eta} =  \cfrac{1}{1+e^{-\eta}}  $$

Where, $\eta = \alpha + \beta_1 X_1 + ... + \beta_n X_n$

In [70]:
eta = np.array(log_reg.coef_[0]*X_train['Gene 10']) + log_reg.intercept_
proba_man = 1/(1+np.exp(-eta))
proba_man 

array([ 0.50381999,  0.51050166,  0.50892305,  0.57011726,  0.51799324,
        0.55112321,  0.48618191,  0.47689209,  0.55040545,  0.50797849,
        0.54348604,  0.55229786,  0.517428  ,  0.59299254,  0.54397856,
        0.54546731,  0.57365142,  0.5818994 ,  0.49445238,  0.56384547,
        0.49944368,  0.48426111,  0.49156325,  0.49276196,  0.48464057,
        0.48457534,  0.51664639,  0.51728775,  0.56089594,  0.51605971,
        0.57962202,  0.57589114,  0.56106776,  0.49718061,  0.5682964 ,
        0.54893586,  0.50156691,  0.58797101,  0.57581275,  0.81265717,
        0.53643663,  0.57573595,  0.5813187 ,  0.59186358,  0.5578684 ,
        0.56464905,  0.58448636,  0.51186126,  0.48185011,  0.56748745,
        0.55013659,  0.57075723,  0.57436776,  0.52712799,  0.49045651,
        0.49077254,  0.51407176,  0.5818815 ,  0.61157109,  0.56121995,
        0.56226807,  0.54644301,  0.54277127,  0.55721174,  0.54834102,
        0.49367883,  0.59992973,  0.57286184])

In [71]:
proba_sklearn = log_reg.predict_proba(X_train_gene10.reshape(-1,1))[:,1]
proba_sklearn

array([ 0.50381999,  0.51050166,  0.50892305,  0.57011726,  0.51799324,
        0.55112321,  0.48618191,  0.47689209,  0.55040545,  0.50797849,
        0.54348604,  0.55229786,  0.517428  ,  0.59299254,  0.54397856,
        0.54546731,  0.57365142,  0.5818994 ,  0.49445238,  0.56384547,
        0.49944368,  0.48426111,  0.49156325,  0.49276196,  0.48464057,
        0.48457534,  0.51664639,  0.51728775,  0.56089594,  0.51605971,
        0.57962202,  0.57589114,  0.56106776,  0.49718061,  0.5682964 ,
        0.54893586,  0.50156691,  0.58797101,  0.57581275,  0.81265717,
        0.53643663,  0.57573595,  0.5813187 ,  0.59186358,  0.5578684 ,
        0.56464905,  0.58448636,  0.51186126,  0.48185011,  0.56748745,
        0.55013659,  0.57075723,  0.57436776,  0.52712799,  0.49045651,
        0.49077254,  0.51407176,  0.5818815 ,  0.61157109,  0.56121995,
        0.56226807,  0.54644301,  0.54277127,  0.55721174,  0.54834102,
        0.49367883,  0.59992973,  0.57286184])

In [76]:
(proba_sklearn == proba_man).sum()

68

In [77]:
np.array_equal(proba_sklearn,proba_man)

True

#### Using all predictor variables in a logistic regression

In [79]:
log_all = LogisticRegression(C=100000, fit_intercept=True)
log_all.fit(X_train_n, y_train)
y_pred_log_all_train = log_all.predict(X_train_n)
y_pred_log_all_test = log_all.predict(X_test_n)
print("Training accuracy for Logistic Regression: ", accuracy_score(y_train, y_pred_log_all_train))
print("Test accuracy for Logistic Regression: ", accuracy_score(y_test, y_pred_log_all_test))

('Training accuracy for Logistic Regression: ', 1.0)
('Test accuracy for Logistic Regression: ', 0.94117647058823528)


In [81]:
log_all.coef_

array([[-0.02457213, -0.04006059, -0.09820024, ...,  0.03252062,
        -0.03102337, -0.11788837]])

In [82]:
log_all.intercept_

array([ 0.00370287])

In [83]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,y_pred_log_all_test)

array([[18,  1],
       [ 1, 14]], dtype=int64)

In [84]:
print(metrics.classification_report(y_test,y_pred_log_all_test))

             precision    recall  f1-score   support

          0       0.95      0.95      0.95        19
          1       0.93      0.93      0.93        15

avg / total       0.94      0.94      0.94        34



#### Incorporating Bootstrapping in a Logistic Regression Model

The main idea behind bootstrapping is to resample our data many times. For each time that we resample our data, we will fit the Logistic Regression model. Then for each resampling, we can find an associated confidence interval. Notice how we have many more predictors than observations. Intuitively, that creates a problem in that we do not have enough observations to "represent" each predictor varaible. Hence, the resampling gives us a way to approximate our standard errors. The goal here is to find confidence intervals for our logistic regression model using ALL the predictors. You may ask, why don't we just fit a logistic regression, then look at the reported standard errors and confidence intervals? The answer is that with more predictors than observations, the reported confidence intervals will look like NaN.   


#### a bootstrap analysis and estimate just how many of our coefficients are significant

In [96]:
NOB = 100 # Number of iterations
boot_coefs = np.zeros((X_train_n.shape[1],NOB)) # Create empty storage array for later use
for i in range(NOB):
    # Sampling WITH replacement the indices of a resampled dataset
    sample_index = np.random.choice(range(len(y_train)), size=len(y_train), replace=True)
    X_train_samples = X_train_n.values[sample_index]
    y_train_samples = y_train.values[sample_index]  
    log_all_boot = LogisticRegression(C=100000, fit_intercept=True)
    log_all_boot.fit(X_train_samples, y_train_samples)
    boot_coefs[:,i] = log_all_boot.coef_
boot_coefs.shape

(6033L, 100L)

If we wanted to find the 95 percent confidence interval, we can use the np.percentile function to find the datapoint that corresponds to the upper and lower bounds. As an example, we have the upper datapoint for the 95% confidence interval. (Why is it 97.5? Think of an interval from 0 to 100, and you want to shade in 95% of that interval, while keeping the length to the left and right of the shading the same. Where does the shading end and begin?)

In [98]:
ci_upper = np.percentile(boot_coefs, 97.5, axis=1)
ci_lower = np.percentile(boot_coefs, 2.5, axis=1)

In [102]:
# ct significant predictors
sig_b_ct = 0
# if ci contains 0, then insignificant
for i in range(len(ci_upper)):
    if ci_upper[i]<0 or ci_lower[i]>0:
        sig_b_ct += 1
print("Significant coefficents at 5pct level = %i / %i" % (sig_b_ct, X_train_n.shape[1]))

Significant coefficents at 5pct level = 1549 / 6033
